### DISCUSSION TITLE
Help with rules for method chain indentation

### DISCUSSION THREAD

**Discussion by mzur** [PHP-CS-Fixer/PHP-CS-Fixer#7319] [Q&A] [answered]
Hi all, I'm in the process of setting up the CS Fixer in my project and have a question regarding method chain indentation. Maybe there is no set of rules that matches the way I did things before or I just couldn't find the correct rules.

Here are examples for the changes that CS Fixer wants to apply to my code:

```diff
         $this->putJson('/api/v1/import/abc123', [
-                'parent_conflicts' => [1 => 'import', 2 => 'existing'],
-            ])
+            'parent_conflicts' => [1 => 'import', 2 => 'existing'],
+        ])
             ->assertStatus(200);
```

```diff
         $parentConflicts = $mergeLabels->filter(function ($label) {
-                return array_key_exists('conflicting_parent_id', $label);
-            })
+            return array_key_exists('conflicting_parent_id', $label);
+        })
             ->each(function ($label) use ($parentConflictResolution) {
```

So basically with a multi-line array or a closure, CS Fixer wants to unindent the part to the same level than the original method call. However, I would like this to be on the same level than the following method in the chain. Can this be accomplished somehow? My configured CS Fixer rules are below.

```php
[
    '@PSR2' => true,
    'ordered_imports' => true,
    'single_space_around_construct' => true,
    'array_indentation' => true,
    'array_syntax' => ['syntax' => 'short'],
    'combine_consecutive_unsets' => true,
    'class_attributes_separation' => ['elements' => ['method' => 'one', ]],
    'no_whitespace_before_comma_in_array' => true,
    'unary_operator_spaces' => true,
    'whitespace_after_comma_in_array' => true,
    'method_chaining_indentation' => true,
]
```

**Comment by Wirone** [comment, marked as answer]
Did you try this?

```php
$this
    ->putJson('/api/v1/import/abc123', [
        'parent_conflicts' => [1 => 'import', 2 => 'existing'],
    ])
    ->assertStatus(200);
```

The problem here is caused by `array_indentation`, not `method_chaining_indentation` itself. Since second argument of `putJson()` is an multiline array, it finds the initial indentation of the line (not opening `[`), which is `$this`', and tries to align the rest to it. If you add new line for the first call in the chain, it works properly:

<img width="1073" alt="image" src="https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/assets/600668/148fc2fe-5a75-42c3-a07e-29df6cbac745">

I am not sure if we should consider this as a bug in `array_indentation` 🤔...

**Reply by mzur** [reply]
Thanks! And do you have an idea what happens with the second example which is not an array?

**Reply by Wirone** [reply]
It's moved by `statement_indentation` with exactly the same reason - it finds initial indentation for `{`, which in fact is taken from `$parentConflicts`' (beginning of the line). You can always print which fixer were applied when you add `--verbose` flag 🙂.

<img width="865" alt="image" src="https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/assets/600668/c422e313-80d5-45cd-b30a-94f27ff7cad4">

You may be interested in [`use_arrow_functions`](https://cs.symfony.com/doc/rules/function_notation/use_arrow_functions.html), because with one-line arrow function it won't be indented incorrectly.

<img width="1126" alt="image" src="https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/assets/600668/cdcdccb2-62f1-4bf9-9d85-3d2f96352fe8">

For multiline closures you would need move `->filter()` to new line.

**Reply by mzur** [reply]
Thanks again! The arrow function would work here but I also have cases that don't fit into an arrow function.

So it looks like both issues would be fixed with a newline after `$this` or the other variable. I could live with this as a coding style in my code base. Is there by any chance a CS Fixer rule for this? :slightly_smiling_face:

**Reply by Wirone** [reply]
I'm afraid we don't have rule for enforcing new line for chained calls. Depending on the amount of affected places in your codebase, I would try running Fixer, then add new line where code was "fixed" and then re-run fixer again, so indentation would become correct. If there are too many occurrences, I would try find & replace based on regex in IDE.

**Reply by mzur** [reply]
Thanks a lot for the quick help!
